File: app/src/main/java/com/savory/login/LoginActivity.java
package com.savory.login;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;

import com.savory.R;
import com.savory.api.clients.savory.SavoryClient;
import com.savory.api.clients.savory.models.SavoryToken;
import com.savory.data.SharedPreferencesClient;
import com.savory.home.HomeActivity;
import com.savory.login.LoginClient.LoginListener;
import com.savory.ui.SimpleBlockingProgressDialog;
import com.savory.utils.UIUtils;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class LoginActivity extends AppCompatActivity implements UserAgreementDialog.Listener {

    @BindView(R.id.facebook_button) FacebookButton facebookButton;

    protected SavoryClient savoryClient;
    private LoginClient loginClient;
    protected SharedPreferencesClient sharedPreferencesClient;

    private UserAgreementDialog userAgreementDialog;
    protected SimpleBlockingProgressDialog progressDialog;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.login);
        ButterKnife.bind(this);

        userAgreementDialog = new UserAgreementDialog(this, this);
        progressDialog = new SimpleBlockingProgressDialog(this, R.string.logging_you_in);

        savoryClient = SavoryClient.get();
        loginClient = new LoginClient(savoryClient);

        sharedPreferencesClient = new SharedPreferencesClient(this);
        String savoryToken = sharedPreferencesClient.retrieveSavoryToken();
        if (savoryToken != null) {
            savoryClient.setSavoryToken(savoryToken);
            startActivity(new Intent(LoginActivity.this, HomeActivity.class));
            finish();
            return;
        }

        loginClient.setListener(new LoginListener() {
            @Override
            public void onLoginPending() {
                progressDialog.show();
            }

            @Override
            public void onSuccessfulLogin(SavoryToken savoryToken) {
                progressDialog.dismiss();

                String token = savoryToken.getToken();
                savoryClient.setSavoryToken(token);
                sharedPreferencesClient.persistSavoryToken(token);
                startActivity(new Intent(LoginActivity.this, HomeActivity.class));
                finish();
            }

            @Override
            public void onLoginCancelled() {}

            @Override
            public void onLoginError() {
                progressDialog.dismiss();
                UIUtils.showLongToast(R.string.login_fail, LoginActivity.this);
            }
        });
    }

    @OnClick(R.id.facebook_button)
    public void onFacebookLoginClicked() {
        userAgreementDialog.show();
    }

    @Override
    public void onUserAgreementAccepted() {
        loginClient.loginWithFacebook(this);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        loginClient.onLoginResult(requestCode, resultCode, data);
    }

    @Override
    public void startActivityForResult(Intent intent, int requestCode) {
        super.startActivityForResult(intent, requestCode);
        overridePendingTransition(R.anim.slide_in_bottom, R.anim.stay);
    }
}

File: app/src/main/java/com/savory/login/LoginActivity.kt
package com.savory.login

import android.content.Intent
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.widget.FrameLayout
import android.widget.Toast
import butterknife.BindView
import butterknife.ButterKnife
import com.facebook.FacebookCallback
import com.facebook.FacebookException
import com.facebook.login.LoginResult
import com.savory.R
import com.savory.api.SavoryClient
import com.savory.api.models.SavoryToken
import com.savory.home.HomeActivity
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class LoginActivity : AppCompatActivity() {

    @BindView(R.id.facebook_button)
    lateinit var facebookButton : FacebookButton

    protected lateinit var savoryClient : SavoryClient

    protected var connectCallback : Call<Response<SavoryToken>>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)
        ButterKnife.bind(this)
        this.savoryClient = SavoryClient.get()

        facebookButton.registerCallback(object : FacebookCallback<LoginResult> {
            override fun onSuccess(result: LoginResult?) {
                Toast.makeText(this@LoginActivity, "We did it!", Toast.LENGTH_SHORT).show()
                savoryClient.connect(result!!.accessToken.token.toString())
                        .enqueue(object : Callback<SavoryToken> {
                            override fun onResponse(call: Call<SavoryToken>?,
                                                    response: Response<SavoryToken>?) {
                                Log.i("LoginActivity", "We did it reddit!")
                                startActivity(Intent(this@LoginActivity, HomeActivity::class.java))
                                finish()
                            }

                            override fun onFailure(call: Call<SavoryToken>?,
                                                   t: Throwable?) {
                                Log.e("LoginActivity", t?.message ?: "Rip")
                            }
                        })
            }

            override fun onCancel() {
            }

            override fun onError(error: FacebookException?) {
            }
        })
    }

    override fun onStop() {
        super.onStop()
        connectCallback?.cancel()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        facebookButton.onFacebookLoginResult(requestCode, resultCode, data)
    }
}
